src/plugins/nestedHeaders/stateManager/headersTree.js

```diff
@@ -0,0 +1,215 @@
+ *   .--(B1)--.           (B2)--.              (B3)--.
+ *  (C1)     (C2)              (C3)                 (C4)
+ *
+ * @type {HeadersTree}
+ */
+/* eslint-enable jsdoc/require-description-complete-sentence */
+export default class HeadersTree {
+  /**
+   * The collection of nested headers settings structured into trees.
+   *
+   * @private
+   * @type {Map<number, TreeNode>}
+   */
+  #rootNodes = new Map();
+  /**
+   * A map that translates the visual column indexes that intersect the range
+   * defined by the header colspan width to the root index.
+   *
+   * @private
+   * @type {Map<number, number>}
+   */
+  #rootsIndex = new Map();
+  /**
+   * The instance of the SourceSettings class.
+   *
+   * @private
+   * @type {SourceSettings}
+   */
+  #sourceSettings = null;
+
+  constructor(sourceSettings) {
+    this.#sourceSettings = sourceSettings;
+  }
+
+  /**
+   * Gets an array of the all root nodes.
+   *
+   * @returns {TreeNode[]}
+   */
+  getRoots() {
+    return Array.from(this.#rootNodes.values());
+  }
+
+  /**
+   * Gets a root node by specified visual column index.
+   *
+   * @param {number} columnIndex A visual column index.
+   * @returns {TreeNode|undefined}
+   */
+  getRootByColumn(columnIndex) {
+    let node;
+
+    if (this.#rootsIndex.has(columnIndex)) {
+      node = this.#rootNodes.get(this.#rootsIndex.get(columnIndex));
+    }
+
+    return node;
+  }
+
+  /**
+   * Gets a tree node by its position in the grid settings.
+   *
+   * @param {number} headerLevel Header level index (there is support only for positive values).
+   * @param {number} columnIndex A visual column index.
+   * @returns {TreeNode|undefined}
+   */
+  getNode(headerLevel, columnIndex) {
+    const rootNode = this.getRootByColumn(columnIndex);
+
+    if (!rootNode) {
+      return;
+    }
+
+    // Normalize the visual column index to a 0-based system for a specific "box" defined
+    // by root node colspan width.
+    const normColumnIndex = columnIndex - this.#rootsIndex.get(columnIndex);
+    let columnCursor = 0;
+    let treeNode;
+
+    // Collect all parent nodes that depend on the collapsed node.
+    rootNode.walkDown((node) => {
+      const { data: { origColspan, headerLevel: nodeHeaderLevel } } = node;
+
+      if (headerLevel === nodeHeaderLevel) {
+        if (normColumnIndex >= columnCursor && normColumnIndex <= columnCursor + origColspan - 1) {
+          treeNode = node;
+
+          return false; // Cancel tree traversing.
+        }
+
+        columnCursor += origColspan;
+      }
+    });
+
+    return treeNode;
+  }
+
+  /**
+   * Builds (or rebuilds if called again) root nodes indexes.
+   */
+  rebuildTreeIndex() {
+    let columnIndex = 0;
+
+    this.#rootsIndex.clear();
+
+    arrayEach(this.#rootNodes, ([, { data: { colspan } }]) => {
+      // Map tree range (colspan range/width) into visual column index of the root node.
+      for (let i = columnIndex; i < columnIndex + colspan; i++) {
```

Comment:
Maybe this `i` name may be more clear?

Reply:
`i` variable is OK here. The variable is encapsulated between brackets that holds only 1 line of code :)